Switch languages without reloading

chrisgarrity opened the pull request:
### Resolves

_What Github issue does this resolve (please include link)?_
#1577 and #1570 I could separate the second into a separate PR, but for now it's here

### Proposed Changes
Creates and loads a combined messages file with all the translations. Switching locales overwrites `Blockly.Msg` table with new values and refreshes the workspace and toolbox. To see it work properly requireed implementing a flag to prevent the toolbox from recycling blocks.

Category labels are still an issue, but this seemed like a good time to get everything so far reviewed. Categories are a bigger can of worms.

### Reason for Changes

Supporting multiple languages

### Test Coverage

Added assertions to the scripts that process message files (js and json) to fail if there are problems. Added `npm run translate` to the test step to make sure messages are ok.

chrisgarrity committed:
* Added `Blockly.ScratchMsgs`
* Added script to create combined locales `Blockly.ScratchMsgs.locales`
* Regenerated en.json after changing single quoted strings in messages.js to double quoted strings.
* Include `create_scratch_msgs.js` in `npm run translate`
locale=<locale-string>` to the URL.
* Added assertions so translate fails if `messages.js` has incorrect strings, or if any of the downloaded locale json files have errors in the translations.
* Playground will load different locales.

chrisgarrity committed:
Switching languages

* pulled locales from transifex for languages that are mostly completely translated
* added option to switch language in playground.
* added missing category labels to `messages.js`

chrisgarrity committed:
Add flag to not recycle flyout blocks when switching languages

Fixes #1570

rachel-fenichel commented on core/blockly.js:
```diff
@@ -62,6 +62,7 @@ goog.require('Blockly.inject');
 goog.require('Blockly.utils');
 goog.require('goog.color');
 
+goog.require('Blockly.ScratchMsgs');
```
Move this between `ggog.require('Blockly.Procedures');` and `goog.require('Blockly.Toolbox');`

rachel-fenichel commented on core/scratch_msgs.js:
```diff
@@ -0,0 +1,44 @@
+ * Name space for the Msg singleton.
+ * Msg gets populated in the message files.
+ */
+goog.provide('Blockly.ScratchMsgs');
```
Nit: swap these two lines.  Imports are ordered alphabetically,

chrisgarrity replied:
@rachel-fenichel really swap? the first is a `provide` not `require`. However, on looking at other files it looks like there's always a blank line between provides and requires, so I'll add that.

rachel-fenichel replied:
You're correct, that's a provide so it's in the right place.  My bad.

rachel-fenichel commented on core/scratch_msgs.js:
```diff
@@ -0,0 +1,44 @@
+Blockly.ScratchMsgs.setLocale = function(locale) {
+  if (Object.keys(Blockly.ScratchMsgs.locales).includes(locale)) {
+    Blockly.Msg = Blockly.ScratchMsgs.locales[locale];
+  }
```
Consider adding an error if the locale string is not recognized.

rachel-fenichel commented on msg/messages.js:
```diff
@@ -24,6 +24,8 @@
  *
  * After modifying this file, run `npm run translate` from the root directory
  * to regenerate `./msg/json/en.json`.
+ * IMPORTANT:
```
This file and the message files are now permanent merge conflicts with blockly.  You'll need to add them to `cleanup.sh` to make sure they're resolved correctly.

Currently `cleanup.sh` just removes (almost) all of the files in `msg/js` and `msg/json`.  You'll need to update it to resolve in favor of the scratch-blocks file.  That code is ~[here](https://github.com/LLK/scratch-blocks/blob/develop/cleanup.sh#L21).

You'll also need to add `msg/messages.js` to the explicit list of files that should always be resolved in favor of scratch-blocks.  That's [here](https://github.com/LLK/scratch-blocks/blob/develop/cleanup.sh#L80).

chrisgarrity replied:
I'm actually debating whether we really want to check in the js and json files. Either way, I'll fix up the cleanup script. Thanks for pointing it out.

rachel-fenichel reviewed (commented):
Looks reasonable.  @AnmAtAnm to confirm that the message tables are being used correctly.

AnmAtAnm commented on i18n/create_scratch_msgs.js:
```diff
@@ -0,0 +1,64 @@
+const enKeys = Object.keys(en).sort().toString();
+
+// Check that translation is valid:
+// elt: array [key, translation] from <locale>.json
```
What does `elt` stand for? Please note in the comment.

AnmAtAnm commented on i18n/create_scratch_msgs.js:
```diff
@@ -0,0 +1,64 @@
+};
+
+const validate = function (json, name) {
+    // this is a little stricter than we need - it would be harmless if the translation had extra keys
```
Actually extra keys have caused real problems in the past, because later keys references do not get replaced. This is the appropriate amount of testing.

AnmAtAnm commented on i18n/create_scratch_msgs.js:
```diff
@@ -0,0 +1,64 @@
+    Object.entries(json).forEach(validatePlaceholders);
+};
+
+let file = '';
```
Good place to use a [multiline template string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals#Multi-line_strings).

AnmAtAnm commented on msg/js/es-419.js:
```diff
@@ -0,0 +1,271 @@
+Blockly.Msg["MOTION_ALIGNSCENE_TOPLEFT"] = "arriba-izquierda";
+Blockly.Msg["MOTION_ALIGNSCENE_TOPRIGHT"] = "arriba-derecha";
+Blockly.Msg["MOTION_XSCROLL"] = "desplazar x";
+Blockly.Msg["MOTION_YSCROLL"] = "desplazar y
```
newline in string leading to unterminated parse. In addition to fixing this value, double check your string escaping in script that generates these files.

chrisgarrity replied:
Good catch - I'll add a check as there is no reason for a newline in any of the msg json files.

AnmAtAnm commented on msg/js/qqq.js:
```diff
@@ -0,0 +1,483 @@
+
+Blockly.Msg["VARIABLES_DEFAULT_NAME"] = "default name - A simple, general default name for a variable, preferably short. For more context, see [[Translating:Blockly#infrequent_message_types]].
+{{Identical|Item}}";
+Blockly.Msg["TODAY"] = "button text - Button that sets a calendar to today's date.
```
More incorrect escaping. (Admittedly, this is just the `qqq` file).

AnmAtAnm commented on msg/json/es-419.json:
```diff
@@ -0,0 +1,265 @@
+    "MOTION_ALIGNSCENE_TOPLEFT": "arriba-izquierda",
+    "MOTION_ALIGNSCENE_TOPRIGHT": "arriba-derecha",
+    "MOTION_XSCROLL": "desplazar x",
+    "MOTION_YSCROLL": "desplazar y\n",
```
The source of the error above.

AnmAtAnm commented on core/scratch_msgs.js:
```diff
@@ -0,0 +1,44 @@
+ */
+Blockly.ScratchMsgs.setLocale = function(locale) {
+  if (Object.keys(Blockly.ScratchMsgs.locales).includes(locale)) {
+    Blockly.Msg = Blockly.ScratchMsgs.locales[locale];
```
I think this will work in most places. It will not work in dropdowns.  See https://github.com/google/blockly/issues/1895

chrisgarrity replied:
@AnmAtAnm is there a particular type of dropdown that shouldn't work? In my vertical playground the looks effect dropdown on the `change %1 effect by %2` updates when I switch languages as does the dropdown on the `go to %1` block. However, I've noticed that we're missing some dropdown strings entirely (previous backdrop), and maybe they will be a problem.

AnmAtAnm replied:
I'm presuming you blocks and static dropdowns use the same JSON-based loading mechanism as Blockly.  Dynamic drop downs and `block.init()` initialized fields should not have this error. There is a possibility your team has fixed the underlying bug, but I haven't heard anything like that. It comes down to how the options table is centralized for the specific dropdown of a particular type of block (so the options can be programmatically updated everywhere), and in that process the option message is dereference once, rather than per-block.

chrisgarrity replied:
I think all our dropdowns are initialized in the `init()` function, so maybe it's moot for scratch-blocks. There are definitely some issues with dynamic dropdowns, but they're actually getting initialized in the gui and that is what I'm working on next.

chrisgarrity committed:
Ignore generated message files

update `.gitignore` to exclude json files from transifex and generated js files.

chrisgarrity committed:
Clarifications and consistency

*  removed all the generated files (<locale>.json, <locale>.js) except for ‘en’ versions and added them to git ignore.
* updated cleanup.sh to use our messages.
* added validation check to fail if any translations include newlines
* skip generating js files other than ‘en’ as we don’t use them
* Revised messages.js to use _single_ quoted strings instead of double (consistent with Blockly - if we need to import more strings), and switch the validations to check for single not double quotes.
* additional small changes for simplicity and ergonomics

Not handling category menu at this time as it’s managed differently in scratch-gui.

kchadha commented on core/blockly.js:
```diff
@@ -62,7 +63,6 @@ goog.require('Blockly.inject');
 goog.require('Blockly.utils');
 goog.require('goog.color');
 
-
```
I think the convention in the rest of the files is to have 2 newlines separating the 'goog.require' statements from the code.

chrisgarrity replied:
👍 no need to introduce unnecessary changes.

kchadha commented on core/flyout_base.js:
```diff
@@ -264,6 +264,14 @@ Blockly.Flyout.prototype.dragAngleRange_ = 70;
 Blockly.Flyout.prototype.scrollAnimationFraction = 0.3;
 
+/**
+ * Whether to recycle blocks when refreshing the flyout. Nothing can be
```
I think this comment needs some clarity, because right now it seems like the second sentence is referring to the initial state of this property, when really it's referring to the fact that this initial value can't be used when the locale changes. I would just take the locale part of it out since it describes a use case for changing this property rather than a description of what this property itself does.

kchadha commented on i18n/create_scratch_msgs.js:
```diff
@@ -0,0 +1,66 @@
+    const placeholdersCount = en[key].match(re) ? en[key].match(re).length : 0;
+    if (placeholdersCount > 0) {
+      const tMatch = translation.match(re);
+      const enMatch = en[key].match(re);
```
`en[key].match(re)` is referenced 3 times here. This `enMatch` declaration can move up above the `placeholdersCount` declaration.

kchadha commented on i18n/create_scratch_msgs.js:
```diff
@@ -0,0 +1,66 @@
+
+let file = `// This file was automatically generated.  Do not modify.
+
+'use strict\';
```
Should this first `'` also be escaped? Alternatively, should none of these be escaped since we're using the `` ` `` formatted strings?

AnmAtAnm replied:
The escape is not necessary inside a template literal (but doesn't hurt). It should be consistent, though.

kchadha commented on i18n/js_to_json.js:
```diff
@@ -12,6 +13,7 @@ const PATH_OUTPUT = path.resolve(__dirname, '../msg/json/en.json');
 const match = function (str) {
     if (str.indexOf('Blockly.Msg.') !== 0) return false;
+    assert.notStrictEqual(str.indexOf("';"), str.length - 2, `[${str}] uses single quoted string, should use double quotes.`);
     if (str.indexOf('";') !== str.length - 2) return false;
```
Should there be a similar assertion for this line? It looks like this will fail silently if " is not the second to last character in the line.

kchadha replied:
Also, a general question, perhaps not directly related to this PR: why not use a regex for this `match` function?

E.g. something like `/^Blockly\.Msg\.*";$/`

chrisgarrity replied:
I think it's for performance. regex is slower and the message strings do have a very predictable format.

kchadha commented on msg/js/en.js:
```diff
@@ -45,6 +45,7 @@ Blockly.Msg["DATA_INDEX_RANDOM"] = "random";
 Blockly.Msg["EVENT_WHENFLAGCLICKED"] = "when %1 clicked";
 Blockly.Msg["EVENT_WHENTHISSPRITECLICKED"] = "when this sprite clicked";
 Blockly.Msg["EVENT_WHENSTAGECLICKED"] = "when stage clicked";
+Blockly.Msg["EVENT_WHENTOUCHINGOBJECT"] = "when this sprite touches %1";
```
Isn't this file being gitignored now? Should these changes be in this PR?

AnmAtAnm replied:
There is a `!` in the modified `.gitignore` for `en.json` and `en.js.`.

kchadha replied:
Oh right! I guess I might have been responding to an offline conversation with @chrisgarrity instead of the actual .gitignore file. Oops. =]

chrisgarrity replied:
The en.{js, json} files have always been checked in, so I just tried to stay consistent.

kchadha commented on msg/json/en.json:
```diff
@@ -39,6 +39,7 @@
     "EVENT_WHENFLAGCLICKED": "when %1 clicked",
     "EVENT_WHENTHISSPRITECLICKED": "when this sprite clicked",
     "EVENT_WHENSTAGECLICKED": "when stage clicked",
+    "EVENT_WHENTOUCHINGOBJECT": "when this sprite touches %1",
```
Same question here as for `msg/js/en.js`

chrisgarrity replied:
We agreed to leave the changes in scratch-blocks for the new event block as Eric is still hoping to get it in before the freeze. This way it'll already be in the translations.

kchadha replied:
Ah, I wasn't actually referring to that particular line, just the same question about the en.json file as with the en.js file, which you've already answered.

kchadha commented on msg/messages.js:
```diff
@@ -261,59 +263,70 @@ Blockly.Msg.SOUND_CHANGEVOLUMEBY = "change volume by %1";
+Blockly.Msg.NEW_VARIABLE = "Make a Variable";
+Blockly.Msg.NEW_VARIABLE_TITLE = "New variable name:";
+Blockly.Msg.VARIABLE_MODAL_TITLE = "New Variable";
+Blockly.Msg.VARIABLE_ALREADY_EXISTS = "A variable named '%1' already exists.";
```
Are we okay with this user visible display message quoting the selected variable name changing to use single quotes? E.g. in this message:

![image](https://user-images.githubusercontent.com/1786240/41314445-c2e2f7c8-6e5a-11e8-93db-97a98dfa9512.png)

AnmAtAnm replied:
Good catch.  Similar quote characters are changed in numerous message string below.

Now I'm curious why we are resolving the quoting consistency problem with `"` strings instead of `'` strings, in this `.js` file.

chrisgarrity replied:
I think it was originally because we're generating json, and that requires double quotes. Given that we mix apostrophe and double quotes in the same string, we're going to be escaping things one way or the other. I'm changing messages.js to require single quotes, but really it could go either way.

kchadha commented on msg/scratch_msgs.js:
```diff
@@ -0,0 +1,8552 @@
+// This file was automatically generated.  Do not modify.
```
If this file is generated, should it be included in this PR?

chrisgarrity replied:
Having this file checked in should allow others to view the blocks in other languages. Otherwise you need credentials for Transifex to download the json files and generate it. In the same way that we check in the en.{js,json} files even though they are also generated.

kchadha commented on package.json:
```diff
@@ -17,9 +17,9 @@
     "test:unit": "node tests/jsunit/test_runner.js",
     "test:lint": "eslint .",
-    "test": "npm run test:lint && npm run test:setup && npm run test:unit",
+    "test": "npm run test:lint && npm run test:setup && npm run test:unit && npm run translate",
```
Maybe the translate step should run before the unit tests (and perhaps before the setup), so that it can fail early for simple failures like using `'` instead of `"`.

kchadha commented on tests/vertical_playground.html:
```diff
@@ -482,6 +496,40 @@ <h1>Vertical Blocks</h1>
           <option value="top">Top</option>
           <option value="bottom">Bottom</option>
         </select>
+        <select name="locale" onchange="setLocale(this.value)">
```
maybe eventually this list should be generated based on what's in `msg/js` or `msg/json`

chrisgarrity replied:
I agree that would be nice to do someday.

kchadha commented on msg/messages.js:
```diff
@@ -24,242 +24,256 @@
  * After modifying this file, run `npm run translate` from the root directory
  * to regenerate `./msg/json/en.json`.
+ * IMPORTANT:
+ * All message strings must use double quotes for the scripts to work properly
```
This comment should be updated now that the strings have changed to use single quotes instead of double quotes.

kchadha reviewed (approved):
LGTM
